### PR TITLE
ARROW-7215: [C++][Gandiva] Implement castVARCHAR(numeric_type) functions

### PR DESCRIPTION
**Pull request by anthonylouisbsb**
This PR implements the castVARCHAR for numeric values inside the Gandiva.

It replaces the logic of the https://github.com/apache/arrow/pull/8158 PR  to change the function output to match the Java language patterns.

### CODE REVIEW

**Comment by anthonylouisbsb** [comment]
@projjal Could you close your old PR https://github.com/apache/arrow/pull/8158?

File: cpp/src/arrow/vendored/double-conversion/double-conversion.cc [hunk 1]
```diff
@@ -84,7 +84,24 @@ void DoubleToStringConverter::CreateExponentialRepresentation(
...
+   * - With EMIT_TRAILING_DECIMAL_POINT disabled -> 0.0009 => 9E-4
+   *
+   * If the mantissa is an integer and the EMIT_TRAILING_ZERO_AFTER_POINT flag is enabled
+   * it will add a '0' character at the end of the mantissa representation. Note that that
+   * flag depends on EMIT_TRAILING_DECIMAL_POINT flag be enabled.*/
+  if(length == 1){
+    if ((flags_ & EMIT_TRAILING_DECIMAL_POINT) != 0) {
+      result_builder->AddCharacter('.');
+    }
+    if ((flags_ & EMIT_TRAILING_ZERO_AFTER_POINT) != 0) {
```
Comment (projjal) [inline cpp/src/arrow/vendored/double-conversion/double-conversion.cc:101]: insert this if statement inside the earlier one

**Review by projjal — COMMENTED** [review]
@pitrou Can you review this. There are changes in arrow/util/formatting and the vendored double-conversion code.

File: cpp/src/arrow/vendored/double-conversion/double-conversion.cc (hunk 1 above)
Comment (anthonylouisbsb) [reply cpp/src/arrow/vendored/double-conversion/double-conversion.cc]: Changes applied.

File: cpp/src/gandiva/formatting_utils.h [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include "arrow/type.h"
+#include "arrow/util/formatting.h"
+#include "arrow/vendored/double-conversion/double-conversion.h"
+
+#ifndef ARROW_SRC_GANDIVA_FORMATTING_UTILS_H_
```
Comment (projjal) [inline cpp/src/gandiva/formatting_utils.h:24]: #pragma once already does the job.

Comment (anthonylouisbsb) [reply cpp/src/gandiva/formatting_utils.h]: Changes applied

**Review by projjal — APPROVED** [review]

**Review by projjal — APPROVED** [review]

**Pull request closed without merging**